Added node chaos scenarios

yashashreesuresh opened the pull request:
This commit:
- Adds a node scenario to stop and start an instance
- Adds a node scenario to terminate an instance
- Adds a node scenario to reboot an instance
- Adds a node scenario to stop the kubelet
- Adds a node scenario to crash the node

Fixes: #8

chaitanyaenr commented on config/config.yaml:
```diff
@@ -4,6 +4,15 @@ kraken:
+            enable_scenario:                               # Set to True to enable the scenario
+            selector:                                      # This enables the nodes with the specified selector to be selected for chaos injection
+            number_of_kills:                               # Number of times to inject the scenario
+        node_termination_scenario:
```
Maybe combine the start_stop scenario and terminate into one since they both need the same details ( selector, number_of_kills )? Something like:
node_scenarios:
    action: start/stop/terminate
    selector:
    number_of_kills:

Thoughts?

chaitanyaenr commented on config/config.yaml:
```diff
@@ -4,6 +4,15 @@ kraken:
+        node_termination_scenario:
+            enable_scenario:
+            selector:
+            number_of_kills:
```
Maybe rename number_of_kills to instance_count?

yashashreesuresh replied:
instance_kill_count?

chaitanyaenr commented on node_scenarios.py:
```diff
@@ -0,0 +1,79 @@
+import kraken.invoke.command as runcommand
+
+
+def node_start_stop_scenario(node_scenario_config):
```
Maybe add a package for node scenarios and import it in the main file i.e run_kraken.py? Thoughts?

yashashreesuresh replied:
Yeah, let's do that.

chaitanyaenr commented on node_scenarios.py:
```diff
@@ -0,0 +1,79 @@
+            client.stop_instances(InstanceIds=[instance_id])
+            instance.wait_until_stopped(InstanceIds=[instance_id])
+            logging.info("Node with instance ID: %s is in stopped state" % (instance_id))
+            time.sleep(30)
```
Let's move the time out to config after changing this to poll instead of sleep. This way the user can set the time out after which the test can fail in case the node is not in the expected state.

chaitanyaenr commented on node_scenarios.py:
```diff
@@ -0,0 +1,79 @@
+import sys
+import time
+import json
+import boto3
```
boto3 module needs to be listed as a dependency by adding it to requirements.txt file.

yashashreesuresh replied:
We get boto3 automatically with AWS CLI installation.

chaitanyaenr replied:
Maybe add awscli to the requirements.txt then?

chaitanyaenr commented on node_scenarios.py:
```diff
@@ -0,0 +1,79 @@
+    selector = node_scenario_config["selector"]
+    number_of_kills = node_scenario_config["number_of_kills"]
+
+    client = boto3.client('ec2')
```
Let's make the boto3 client definition global as we use it in other functions related to the node scenarios as well.

chaitanyaenr commented on node_scenarios/node_start_stop_scenario.yml:
```diff
@@ -0,0 +1,3 @@
+label_selector:                                              # This enables the nodes with the specified label selector to be selected for chaos injection
+instance_kill_count:                                         # Number of times to inject the scenario in the nodes
+timeout:                                                     # Set the time out after which the test can fail in case the node is not in the expected state after chaos scenario injection
```
Think the config should accept a list of node scenarios ( start, stop, restart, terminate ) similar to how powerfulseal does it - https://github.com/bloomberg/powerfulseal/blob/master/tests/policy/example_config2.yml except that we will ask user for list of actions ( start, stop, restart, terminate ) instead of ip address's/hostnames. Thoughts?

mffiedler commented on kraken/node_actions/node_scenarios.py:
```diff
@@ -0,0 +1,117 @@
+import time
+import json
+import yaml
+import boto3
```
Can we create an abstract node scenarios class and have cloud specific implementing classes for it that use the cloud specific libraries like boto3 (e.g. aws_node_scenarios)?  The config can contain an option for the cloud type to use and the client can just deal with generic scenarios like start/stop, reboot, termination.

yashashreesuresh replied:
Yeah, this sounds good. I will make the changes to implement the same. Thank you for the feedback! @mffiedler

chaitanyaenr commented on kraken/node_actions/aws_node_scenarios.py:
```diff
@@ -0,0 +1,177 @@
+                sys.exit(1)
+
+
+    def node_start_stop_scenario(self):
```
The node_start_scenario and stop functions can be reused by passing the node_name/selector to act on instead of repeating the logic. Thoughts?

chaitanyaenr commented on node_scenarios/node_scenarios_example.yml:
```diff
@@ -0,0 +1,16 @@
+    - node_reboot_scenario
+    node_name:
+    label_selector: node-role.kubernetes.io/infra
+    instance_kill_count: 1
```
Maybe rename instance_kill_count to instance_count since the action can also be a reboot/stop?

paigerube14 commented on node_scenarios/node_scenarios_example.yml:
```diff
@@ -0,0 +1,16 @@
+node_scenarios:
+  - actions: 
```
It might be nice to add a name to each node scenario here, similar to pod kill scenarios

yashashreesuresh replied:
The items under `actions` refer to the node scenarios to be applied on a selected node.

paigerube14 replied:
Sorry, I meant above that 

node_scenarios:
  - name: Stop and Start Scenario
    actions: 
    - node_stop_scenario
    - node_start_scenario
    node_name:
    label_selector: node-role.kubernetes.io/worker
    instance_kill_count: 1
    timeout: 120
    cloud_type: aws
  - name: Reboot Scenario
    actions:
    - node_reboot_scenario
    node_name:
    label_selector: node-role.kubernetes.io/infra
    instance_kill_count: 1
    timeout: 120
    cloud_type: aws

yashashreesuresh replied:
In the above case, if 5 different scenarios have to be applied on the same node one after another, the same configurations have to be specified 5 times which is redundant.

paigerube14 replied:
I think my location of this comment is making it confusing and I think we are talking about two different things here. I am just suggesting that each node scenario in the list has a name, not each action and no need for redundancy 

Here is an example from the etcd scenario. Here the scenario itself has a name and then you can perform multiple steps/actions
```
scenarios:
  **- name: "delete etcd pods"**
    steps:
    - podAction:
        matches:
          - labels:
              namespace: "openshift-etcd"
              selector: "k8s-app=etcd"
```


Similarly, I am suggesting a name to the top of each node scenario you want to perform on a node. 
```
node_scenarios:
  **- name: Stop and Start Scenario**
    actions: 
    - node_stop_scenario
    - node_start_scenario
    node_name:
    label_selector: node-role.kubernetes.io/worker
    instance_kill_count: 1
    timeout: 120
    cloud_type: aws
```

yashashreesuresh replied:
Suppose we want to perform stop, start, terminate, restart actions on node X. The actions would include node_stop_scenario, node_start_scenario, node_terminate_scenario, node_restart_scenario. In this case would the name be Start, Stop, Terminate and Restart scenario?

paigerube14 replied:
Yeh I see your point now, I agree it's redundant

yashashreesuresh commented:
The current implementation uses abstract class; class for each cloud type inherits the abstract class. With this, the node scenario functions can to be called only after creating an object of the class. The advantage with using abstract class is that it makes sure all the cloud types support all the scenarios. But I think it would be easier to directly call functions without creating an object. Thoughts? @mffiedler @paigerube14

mffiedler commented:
I think the use of the abstract class to allow for cloud-specific scenario implementations where the cloud instance APIs are required is the correct approach.    We also need to allow for scenarios which are not cloud specific (see the node crash scenarios which use oc debug node in the PR @paigerube14 started in #18 for an example).   Where possible, having a default/common implementation of the scenario would be beneficial and it can be overridden by a cloud-specific implementation where needed.   Hope that makes sense.

paigerube14 commented:
I completely agree with Mike, I think the abstract might actually end up being easier for all the cloud providers to have a common layout of function. 

@yashashreesuresh did you want me to add my 2  functions (kubelet_action and crash_node) for the oc debug parts that Mike is talking about to your branch? How did you want to do that?

yashashreesuresh commented:
> I completely agree with Mike, I think the abstract might actually end up being easier for all the cloud providers to have a common layout.
> 
> @yashashreesuresh did you want me to add my 2 functions (kubelet_action and crash_node) for the oc debug parts that Mike is talking about to your branch? How did you want to do that?

@paigerube14 I have added the kubelet scenario as https://github.com/openshift-scale/kraken/pull/10/files#diff-d7c8b18dc1f8fd23bc1e62bdee0ccdceL8 subprocess.check_output doesn't finish to completion because once the kubelet is stopped, debug pod never terminates. Therefore, I had to use https://github.com/openshift-scale/kraken/pull/10/files#diff-d7c8b18dc1f8fd23bc1e62bdee0ccdceR15 subprocess.run command. 
I was trying your node_crash scenario and I observed many times the command gets executed but the node status doesn't change from Ready->NotReady->Ready. And sometimes the command pauses and doesn't finish to completion and I have to use Ctrl+C. However, I have added your command here https://github.com/openshift-scale/kraken/pull/10/files#diff-54651a20a8f57ef1c067dbaac8fdc0a6R63. It needs to be changed.

yashashreesuresh commented:
All the scenarios are added, PTAL @chaitanyaenr

paigerube14 commented:
> > I completely agree with Mike, I think the abstract might actually end up being easier for all the cloud providers to have a common layout.
> > @yashashreesuresh did you want me to add my 2 functions (kubelet_action and crash_node) for the oc debug parts that Mike is talking about to your branch? How did you want to do that?
> 
> @paigerube14 I have added the kubelet scenario as https://github.com/openshift-scale/kraken/pull/10/files#diff-d7c8b18dc1f8fd23bc1e62bdee0ccdceL8 subprocess.check_output doesn't finish to completion because once the kubelet is stopped, debug pod never terminates. Therefore, I had to use https://github.com/openshift-scale/kraken/pull/10/files#diff-d7c8b18dc1f8fd23bc1e62bdee0ccdceR15 subprocess.run command.
> I was trying your node_crash scenario and I observed many times the command gets executed but the node status doesn't change from Ready->NotReady->Ready. And sometimes the command pauses and doesn't finish to completion and I had to use Ctrl+C. However, I have added your command here https://github.com/openshift-scale/kraken/pull/10/files#diff-54651a20a8f57ef1c067dbaac8fdc0a6R63. It needs to be changed.

I did have to use subprocess.Popen to get the stop kubelet to return properly. I'm not sure about the node_crash scenario. I'm not sure if the node will go to the notReady state. I was going off of the commands Mike had laid out in the original issue. I think for the node_crash scenario I also had to use subprocess.Popen to get a proper response. I think I thought either subprocess function worked so I didn't add it to my pull request, sorry about that. I tried the node_crash scenario from my branch and the Popen properly returns

yashashreesuresh commented:
> I did have to use subprocess.Popen to get the stop kubelet to return properly. I'm not sure about the node_crash scenario. I'm not sure if the node will go to the notReady state. I was going off of the commands Mike had laid out in the original issue. I think for the node_crash scenario I also had to use subprocess.Popen to get a proper response. I think I thought either subprocess function worked so I didn't add it to my pull request, sorry about that. I tried the node_crash scenario from my branch and the Popen properly returns

I tried with Popen, it's not working as I expected. The subprocess.Popen("oc debug node/" + node + " -- chroot /host dd if=/dev/urandom of=/proc/sysrq-trigger") command finishes. Only after few seconds the node goes to NotReady state. I expected the node to go to NotReady state while the Popen command is in execution because the node when to NotReady state after the debug pod is removed. Few times, the node state doesn't change at all.
I am not sure what's happening.

Anyways, I have removed the node_crash_scenario. You can add your code here https://github.com/openshift-scale/kraken/pull/10/files#diff-54651a20a8f57ef1c067dbaac8fdc0a6R58 once this PR is merged as it worked right for you.

yashashreesuresh commented:
All the changes are made in the code. PTAL, @chaitanyaenr @paigerube14 @mffiedler

mffiedler commented:
For node crash scenarios, nodes will not show immediately as NotReady in the api.   They have to miss a readiness check interval from the API server perspective.    If a node is in trouble with memory or disk pressure but still running, it can self report the issue immediately.  But for a crashed node, the cluster needs to notice it is gone to mark it NotReady.

yashashreesuresh commented:
> For node crash scenarios, nodes will not show immediately as NotReady in the api. They have to miss a readiness check interval from the API server perspective. If a node is in trouble with memory or disk pressure but still running, it can self report the issue immediately. But for a crashed node, the cluster needs to notice it is gone to mark it NotReady.

Added back the node crash scenario that was removed.

paigerube14 commented on config/config.yaml:
```diff
@@ -4,6 +4,8 @@ kraken:
         -    scenarios/openshift-kube-apiserver.yml                           
         -    scenarios/openshift-apiserver.yml
+    node_scenarios:                                        # List of chaos node scenarios to load
+        -   scenarios/node_scenarios_example.yml
```
It might be also nice to show an example of the node_crash and stop_kubelet in a separate scenario or also in this one

paigerube14 replied:
In addition, it could be nice to maybe have a separate README defining how to set up a node scenario and all the possible actions that you can have.

yashashreesuresh replied:
I have made changes in the README, PTAL.

paigerube14 replied:
LGTM thanks for adding that!

paigerube14 commented:
Is it expected that we set up the boto3 region and some other attributes before running? I did not set up of boto3 and I am getting an error` botocore.exceptions.NoRegionError: You must specify a region.` This might be good to add to the README or it might be nice to either specify those in the config or in the scenario itself where you define the cloud type to aws

yashashreesuresh commented:
> Is it expected that we set up the boto3 region and some other attributes before running? I did not set up of boto3 and I am getting an error` botocore.exceptions.NoRegionError: You must specify a region.` This might be good to add to the README or it might be nice to either specify those in the config or in the scenario itself where you define the cloud type to aws

boto3 client is created in the `default` region specified in the `~/.aws/config` file. This file is created when awscli is installed. Which version of AWS CLI are use using?

paigerube14 commented:
> > Is it expected that we set up the boto3 region and some other attributes before running? I did not set up of boto3 and I am getting an error` botocore.exceptions.NoRegionError: You must specify a region.` This might be good to add to the README or it might be nice to either specify those in the config or in the scenario itself where you define the cloud type to aws
> 
> boto3 client is created in the `default` region specified in the `~/.aws/config` file. This file is created when awscli is installed. Which version of AWS CLI are use using?

Ah I didn't have an aws cli installed. I definitely think this should be put in the read me as an extra set up steps because that is not clear. Maybe just specify if using the specific node scenarios you'll need to set up aws cli with certain config. I followed https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html if you want to use this in the documentation

yashashreesuresh commented:
> Ah I didn't have an aws cli installed. I definitely think this should be put in the read me as an extra set up steps because that is not clear. Maybe just specify if using the specific node scenarios you'll need to set up aws cli with certain config. I followed https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html if you want to use this in the documentation

Yeah, I will add it in the README. Thank you!

mffiedler commented:
/hold
I need some time to review and test this.  It has gotten pretty big.   In the future I have the suggestion that we add 1 scenario per PR.

yashashreesuresh commented:
> /hold
> I need some time to review and test this. It has gotten pretty big. In the future I have the suggestion that we add 1 scenario per PR.

Sure.

chaitanyaenr commented on README.md:
```diff
@@ -56,6 +58,43 @@ The report is generated in the run directory and it contains the information abo
 [Cerberus](https://github.com/openshift-scale/cerberus) can be used to monitor the cluster under test and the aggregated go/no-go signal generated by it can be consumed by Kraken to determine pass/fail. This is to make sure the Kubernetes/OpenShift environments are healthy on a cluster level instead of just the targeted components level. It is highly recommended to turn on the Cerberus health che ...
 
+### Kubernetes/OpenShift node chaos scenarios supported
+Following node chaos scenarios are supported:
```
We might want to mention that the cloud related scenarios like start/stop/reboot are supported only for AWS as of now.

yashashreesuresh replied:
Added it.

chaitanyaenr commented on kraken/node_actions/abstract_node_scenarios.py:
```diff
@@ -0,0 +1,68 @@
+        self.node_reboot_scenario(instance_kill_count, node, timeout)
+        logging.info("stop_start_kubelet_scenario has been successfully injected!")
+
+    # Node scenario to crash the node
```
We might want to add a note in the readme that the node needs to be rebooted to recover from node crash and that it can be added as an action in the node scenarios config.

yashashreesuresh replied:
Added it.

chaitanyaenr commented on README.md:
```diff
@@ -56,6 +58,43 @@ The report is generated in the run directory and it contains the information abo
+
+**NOTE**: With aws as the cloud type, make sure [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) is installed.
+
 ### Kubernetes/OpenShift chaos scenarios supported
```
We might want to move the node scenarios under "Kubernetes/OpenShift chaos scenarios supported" section.

yashashreesuresh replied:
Done.

chaitanyaenr commented:
@paigerube14 @mffiedler Thoughts?

paigerube14 commented on README.md:
````diff
@@ -57,7 +59,51 @@ The report is generated in the run directory and it contains the information abo
+
+
+```
+node_scenarios:
````
Should we make this in a separate readme for the node scenarios?

yashashreesuresh replied:
I think it would be better to place everything in the main Readme as the user gets to know about different scenarios supported by Kraken.

mffiedler commented:
I think we should merge this and do any follow up work needed as a separate PR

LGTM
